server/aesdsocket.c

```diff
@@ -58,23 +132,222 @@ static void signal_handler (int signo)
+	if(thread_func_args->transmit_buffer == NULL)
+	{
+		close(thread_func_args->client_fd);
+		printf("Error while doing realloc\n");
+		close_all();
+		exit (EXIT_FAILURE);
+	}
+
+	rc = pthread_mutex_lock(&main_mutex); //lock mutex
+	if(rc != 0)
+	{
+		perror("Mutex lock failed\n");
+		close_all();
+		exit (EXIT_FAILURE);
+	}
+
+	// block the signals SIGINT and SIGTERM
+	rc = sigprocmask(SIG_BLOCK, &thread_func_args->signal_mask, NULL);
+	if(rc == -1)
+	{
+		printf("Error while blocking the signals SIGINT, SIGTERM");
+		close_all();
+		exit (EXIT_FAILURE);
+	}
+
+	lseek(write_fd, 0, SEEK_SET); //set cursor to start
+	read_bytes = read(write_fd, thread_func_args->transmit_buffer, size); //read bytes
+	if(read_bytes < 0)
+	{
+		close(thread_func_args->client_fd);
+		printf("Error in reading bytes\n");
+		close_all();
+		exit (EXIT_FAILURE);
+	}
+
+    send_bytes = send(thread_func_args->client_fd, thread_func_args->transmit_buffer, read_bytes, 0);//send bytes
+    if(send_bytes < 0)
+    {
+	close(thread_func_args->client_fd);
+	printf("Error in sending bytes\n");
+	close_all();
+	exit (EXIT_FAILURE);
+    }
+    rc = pthread_mutex_unlock(&main_mutex); // unlock mutex
+    if(rc != 0)
+    {
+	perror("Mutex unlock failed\n");
+	close_all();
+	exit (EXIT_FAILURE);
+    }
+    // unblock the signals SIGINT and SIGTERM
+    rc = sigprocmask(SIG_UNBLOCK, &thread_func_args->signal_mask, NULL);
+    if(rc == -1)
+    {
+	printf("Error while unblocking the signals SIGINT, SIGTERM");
+	close_all();
+	exit (EXIT_FAILURE);
+    }
+    // close connection
+    close(thread_func_args->client_fd);
+    //syslog(LOG_DEBUG, "Closed connection from %s", ipstr);
+    free(thread_func_args->buffer);
+    free(thread_func_args->transmit_buffer);
+    thread_func_args->thread_complete_success = true;
+
+}
+
+//add timespecs and store in result
+static void timespec_add(const struct timespec *ts_a, const struct timespec *ts_b, struct timespec *result)
+{
+    result->tv_sec = ts_a->tv_sec + ts_b->tv_sec;
+    result->tv_nsec = ts_a->tv_nsec + ts_b->tv_nsec;
+    if( result->tv_nsec > 1000000000L )
+    {
+       	result->tv_nsec -= 1000000000L;
+        result->tv_sec ++;
+    }
+}
+
+
+
+//service every 10 seconds
+static void timer10sec_thread(union sigval sigval)
 {
+    char time_buffer[TIMER_BUFFER_SIZE];
+    time_t current_time;
+    struct tm *timer_info;
+    time(&current_time);
+    timer_info = localtime(&current_time);
 
+    int timer_buffer_size = strftime(time_buffer,TIMER_BUFFER_SIZE,"timestamp:%a, %d %b %Y %T %z\n",timer_info);
 
+    pthread_mutex_lock(&main_mutex);
 
-    sigset_t set;
-    sigemptyset(&set);           // empty the set
+    // write to file
+    int timer_writebytes = write(write_fd,time_buffer,timer_buffer_size);
+    if(timer_writebytes == -1)
+    {
+        printf("Error in writing time to file\n");
+        close_all();
+        exit(-1);
```

Comment:
@Sharath-Jonnala exiting the entire process due to this failure is a little harsh.  The error handler could be an issue if this line were a `return` and happened when you were holding the mutex.